File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.py
class Solution:
    def topKFrequent(self, nums, k):

        # Step 1: Count frequency
        freq = {}
        for num in nums:
            if num in freq:
                freq[num] += 1
            else:
                freq[num] = 1

        result = []

        # Step 2: Find top k frequent elements
        for _ in range(k):

            max_freq = -1
            max_key = None

            # find element with maximum frequency
            for key in freq:
                if freq[key] > max_freq:
                    max_freq = freq[key]
                    max_key = key

            result.append(max_key)

            # remove selected element
            del freq[max_key]

        return result

File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {

        List<List<Integer>> result = new ArrayList<>();
        List<Integer> sum = new ArrayList<>();

        backtrack(nums, target, sum, 0, result);

        return result;
    }

    public void backtrack(int[] nums, int target, List<Integer> sum, int i, List<List<Integer>> result) {

        if (target == 0) {
            result.add(new ArrayList<>(sum));
            return;
        }

        if (target < 0 || i == nums.length) {
            return;
        }

        sum.add(nums[i]);
        backtrack(nums, target - nums[i], sum, i, result);

        sum.remove(sum.size() - 1);

        backtrack(nums, target, sum, i + 1, result);
    }
}